### ISSUE TITLE
Handle to modify the knitted document header

### ISSUE THREAD

**Issue by aoles** [rstudio/rmarkdown#888]
I find the following behavior potentially limiting: the `metadata` argument to `post_knit`, `pre_processor` and `post_processor` functions passed to `output_format` contains raw values from document YAML front matter *before* they are parsed by **knitr**, even though conceptually these functions are executed after knitting the document.

In my use case I have a metadata field containing an R expression, which I would like to access after it has been evaluated. In fact, I didn't find a way to get the knitted YAML metadata in the current API. What I would like to ultimately achieve is to process the *knitted* header in order to alter some of its fields before the document is handed over to pandoc.

I came up with a [`pre_processor` function](https://github.com/Bioconductor/BiocStyle/blob/49f727c6f4614d127334dcd69484a05757561680/R/rmd_common.R) which extracts the metadata header from the knitted md file by using `rmarkdown:::partition_yaml_front_matter`. But maybe there is a better way of modifying the document's front matter without the need of resorting to **rmarkdown**'s internal functions?

In summary, it would be great if you would consider providing ways to access the knitted document metadata, and to modify the document's front matter.

Thanks for your help, and keep up the good work!
Andrzej

**Comment by yihui** [comment]
I think we can probably export a helper function that calls `partition_yaml_front_matter()` and `yaml::yaml.load()`, but on the other hand, I think this is trivially easy for you to do it by yourself without relying on any internal rmarkdown functions:

```r
read_yaml_meta = function(file) {
  x = readLines(file, encoding = 'UTF-8')
  i = grep('^---\\s*$', x)
  if (length(i) < 2 || i[2] - i[1] < 2) return()
  meta = paste(x[(i[1] + 1):(i[2] - 1)], collapse = '\n')
  yaml::yaml.load(meta)
}
```

**Comment by aoles** [comment]
Many thanks you for your answer!

Sure, I could extract the metadata manually with the help of the of the suggested code. But then, my use case is actually slightly more complicated, because I would like to not only load the knitted matadata, but also swap the original yaml header by the processed one. Therefore I need both the document front matted and the document body in order to put them back together afterwards. Please have a look at [my code](https://github.com/Bioconductor/BiocStyle/blob/49f727c6f4614d127334dcd69484a05757561680/R/rmd_common.R) to get an idea of what I'm trying to achieve. So by the end of the day I would actually need to replicate the functionality of partitioning the document into front matter and body, which is already internally implemented in **rmarkdown**.

I would really appreciate it if you consider exporting a helper function, possibly `partition_yaml_front_matter()` itself, which would provide an easy and robust way of accessing document's front matter and body.

Cheers,
Andrzej
